File: tests/main_test.go
package main

import (
	"context"
	"os"
	"testing"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"github.com/nicolasgomezaragon/FinancialAPI/pkg/models"
	"github.com/nicolasgomezaragon/FinancialAPI/pkg/utils"
	"github.com/nicolasgomezaragon/FinancialAPI/pkg/db"
)

func TestReadToken(t *testing.T) {
	// Create a temporary token file
	file, err := os.CreateTemp("", "token.txt")
	if err != nil {
		t.Fatalf("Failed to create temp file: %s", err)
	}
	defer os.Remove(file.Name())

	// Write a test token to the file
	token := "test_token"
	if _, err := file.WriteString(token); err != nil {
		t.Fatalf("Failed to write token: %s", err)
	}
	file.Close()

	// Test readToken function
	readToken, err := utils.ReadToken(file.Name())
	if err != nil {
		t.Fatalf("Failed to read token: %s", err)
	}

	if readToken != token {
		t.Errorf("Expected %s, got %s", token, readToken)
	}
}

func TestSaveToMongoDB(t *testing.T) {
	// Mock data
	data := map[string]models.DailyData{
		"2025-03-31": {Open: "100", High: "110", Low: "90", Close: "105", Volume: "1000"},
	}

	// Connect to MongoDB
	client, err := mongo.Connect(context.TODO(), options.Client().ApplyURI("mongodb://localhost"))
	if err != nil {
		t.Fatalf("Failed to connect to MongoDB: %s", err)
	}
	defer client.Disconnect(context.TODO())

	// Use a test database and collection
	dbName := "test_db"
	collectionName := "test_collection"
	collection := client.Database(dbName).Collection(collectionName)

	// Clean up after test
	defer client.Database(dbName).Drop(context.TODO())

	// Test saveToMongoDB function
	err = db.SaveToMongoDB(data, dbName, collectionName)
	if err != nil {
		t.Fatalf("Failed to save to MongoDB: %s", err)
	}

	// Verify data was saved
	var result struct {
		Open  string `bson:"open"`
		Close string `bson:"close"`
	}
	err = collection.FindOne(context.TODO(), bson.M{"date": "2025-03-31"}).Decode(&result)
	if err != nil {
		t.Fatalf("Failed to find data: %s", err)
	}

	if result.Open != "100" || result.Close != "105" {
		t.Errorf("Expected Open: 100, Close: 105, got Open: %s, Close: %s", result.Open, result.Close)
	}
}

File: tests/main_test.py
import os
import tempfile
import unittest
from pymongo import MongoClient
from pkg.utils import read_token
from pkg.db import save_to_mongo_db
from pkg.models import DailyData

class TestUtils(unittest.TestCase):
    def test_read_token(self):
        # Create a temporary token file
        with tempfile.NamedTemporaryFile(delete=False) as temp_file:
            temp_file.write(b"test_token")
            temp_file_name = temp_file.name
        
        # Test read_token function
        token, err = read_token(temp_file_name)
        self.assertIsNone(err)
        self.assertEqual(token, "test_token")
        
        # Clean up
        os.remove(temp_file_name)

class TestDB(unittest.TestCase):
    def test_save_to_mongo_db(self):
        # Mock data
        data = {
            "2025-03-31": DailyData(open="100", high="110", low="90", close="105", volume="1000")
        }
        
        # Connect to MongoDB
        client = MongoClient('localhost', 27017)
        db_name = "test_db"
        collection_name = "test_collection"
        db = client[db_name]
        collection = db[collection_name]
        
        # Clean up after test
        self.addCleanup(client.drop_database, db_name)
        
        # Test save_to_mongo_db function
        err = save_to_mongo_db(data, db_name, collection_name)
        self.assertIsNone(err)
        
        # Verify data was saved
        result = collection.find_one({"date": "2025-03-31"})
        self.assertIsNotNone(result)
        self.assertEqual(result["open"], "100")
        self.assertEqual(result["close"], "105")

if __name__ == '__main__':
    unittest.main()
